### ISSUE TITLE
Code coverage

### ISSUE THREAD

**Issue by 3d4m-vladimir** [emscripten-core/emscripten#13046]
Is there anything done in emscripten for enabling Tests Code Coverage reports generation?

I found [this very interesting article](https://motley-coder.com/2019/03/31/webassembly-coverage/) on how to do it manually.
Requires to recompile **compiler-rt**.

**Comment by sbc100** [comment]
I don't think we have any direct support for that today, but its sounds like it would be fairly easy to add. 

Sounds like it would involve a new option and new compiler-rt variant added to `tools/system_libs.py`

**Comment by 3d4m-vladimir** [comment]
Is this the place to start looking at https://github.com/emscripten-core/emscripten/blob/master/tools/system_libs.py#L642

I never dealt with the internals of emscripten. Probably will have lots of questions.

**Comment by sbc100** [comment]
I took an initial stab at getting this to work:

https://github.com/emscripten-core/emscripten/pull/13072

It works up until llvm-cov fails to parse the data sections in the wasm executable.

**Comment by emaxx-google** [comment]
Hello, is there any update on this effort?
(We're looking into ways to set up coverage analysis for Google Smart Card Connector, and IIUC currently the only workaround is to switch to the regular clang/gcc, but this doesn't let test the code in the WebAssembly environment.)

**Comment by AnnieXia** [comment]
`emcc -fprofile-instr-generate -fcoverage-mapping demo.cpp`

wasm-ld: error: /var/folders/61/x35v93n50_5g1dfjrtprr2dw0000gn/T/emscripten_temp_xjrzk63_/demo_0.o: undefined symbol: __llvm_profile_runtime

have emcc supported code coverage? need help :)

**Comment by sbc100** [comment]
No, emcc does not yet support this method of doing code coverage.   #13072 is a start, but that has yet to land.

**Comment by junyuecao** [comment]
It looks like [SanitizerCoverage](https://clang.llvm.org/docs/SanitizerCoverage.html) works. I've tried the demo by replacing clang++ to em++
```
em++ -g  -fsanitize-coverage=trace-pc-guard trace-pc-guard-example.cc -c
em++ trace-pc-guard-cb.cc trace-pc-guard-example.o -fsanitize=address -sNO_EXIT_RUNTIME -o trace-pc-guard-cb.html
```

**Comment by 3d4m-vladimir** [comment]
@junyuecao this is not the same.

**Comment by bbarwik** [comment]
Hey. I've created a tutorial how to generate code coverage for webassembly projects. This one is dedicated for blockchains using WASM VM, however I think it may be useful to you too - https://github.com/hknio/code-coverage-for-webassembly

**Comment by nihui** [comment]
Hi

I implemented the required routines for wasm code coverage, see https://github.com/Tencent/ncnn/pull/5903

I can use node to run the wasm program, collect gcda data, and combine it with the llvm-cov gcov wrapper to get a reliable coverage report.

It does not require compiler-rt library, you could just add my simplegcov.cpp into your project build, and compile with some special flags for coverage working


sample usage guide
https://github.com/Tencent/ncnn/issues/5902#issuecomment-2655366538

wasm code coverage

```shell
# compile with --coverage -fprofile-arcs -ftest-coverage -g
emcc test.cpp -c -o test.o --coverage -fprofile-arcs -ftest-coverage -g
emcc simplegcov.cpp -c -o simplegcov.o --coverage -fprofile-arcs -ftest-coverage -g

# link with --coverage -g -s FORCE_FILESYSTEM=1 -sEXIT_RUNTIME=1
# link with -lnodefs.js -lnoderawfs.js
emcc test.o simplegcov.o -o test.js -s EXPORTED_FUNCTIONS='["_main"]' --coverage -g -s FORCE_FILESYSTEM=1 -sEXIT_RUNTIME=1 -lnodefs.js -lnoderawfs.js

# run wasm
node test.js

# collect gcda and generate output html
lcov -d ./ -c --gcov-tool `pwd`/llvm-gcov -o lcov.info --ignore-errors inconsistent
genhtml lcov.info -o ./output --ignore-errors inconsistent
```

https://github.com/Tencent/ncnn/pull/5903

the llvm-gcov wrapper
```
#!/bin/sh
exec llvm-cov gcov "$@"
```
